src/ddmd/expressionsem.d

```diff
@@ -8771,6 +8808,100 @@ Expression binSemanticProp(BinExp e, Scope* sc)
     return null;
 }
 
+/********************************************************************************
+ * Helper function to consolidate common code for visit(BinAssignExp)
+ * visit(PowAssignExp), and visit(CatAssignExp).  It rewrites expressions
+ * of the form, for example, `s.x += 1` to `s.x(s.x() += 1)` if s.x is
+ * a getter and setter `@property`.
+ * Params:
+ *      e = the expression to rewrite
+ *      sc = the semantic scope
+ * Returns:
+ *      the rewritten expression if the procedure succeeds, an `ErrorExp` if the
+ *      and error is encountered, or `null` if `e.e1` is not a `DotIdExp`.
+ */
+Expression binAssignRewriteProp(BinAssignExp e, Scope* sc)
+{
+    // https://issues.dlang.org/show_bug.cgi?id=8006
+
+    if (e.e1.op == TOKdotid)
+    {
+        // Consider s.x += 1;
+        // We want to rewrite that expression to s.x(s.x() + 1);
+        // However, resolvePropertiesX will change the type of s.x() to void,
+        // so we first need to copy the rhs s.x() to a temporary.
+        // Utimately, we will get something like the following:
+        // int __tmp = s.x();
+        // s.x(__tmp + 1);
+        //********************************************************************
+
+        // Need e1 for the call to resolvePropertiesX
+        Expression e1 = e.e1.semantic(sc);
+        if (e1.op == TOKerror)
+        {
+            return e1;
+        }
+
+        // int __tmp = s.x();
+        auto v = copyToTemp(0, "__tmp", e.e1);
+        auto de = new DeclarationExp(e.e1.loc, v);
+        auto ve = new VarExp(e.e1.loc, v);
+        Expression ea = Expression.combine(de, ve);
```

Comment:
If you pass a copy of e1 to `resolvePropertiesX`, then this shouldn't be needed.  Codegen looks better at least without this.